hpx/runtime/serialization/detail/serialize_collection.hpp

```diff
@@ -0,0 +1,122 @@
+//  Copyright (c) 2017 Anton Bikineev
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef HPX_SERIALIZATION_DETAIL_SERIALIZE_COLLECTION_HPP
+#define HPX_SERIALIZATION_DETAIL_SERIALIZE_COLLECTION_HPP
+
+#include <hpx/config.hpp>
+
+#include <type_traits>
+
+namespace hpx { namespace serialization { namespace detail {
+
+    // default fallbacks. in fact, they should never be called
+    template <class Archive, class T>
+    void save_construct_data(Archive&, T*, unsigned)
+    {
+    }
+
+    template <class Archive, class T>
+    void load_construct_data(Archive&, T* t, unsigned)
+    {
+        ::new (t) T;
+    }
+
+    template <class Value>
+    class save_collection_impl
+    {
+        struct default_
+        {
+            template <class Archive, class Collection>
+            static void call(Archive& ar, const Collection& collection)
+            {
+                for (const auto& i: collection)
+                    ar << i;
+            }
+        };
+
+        struct non_default_
+        {
+            template <class Archive, class Collection>
+            static void call(Archive& ar, const Collection& collection)
+            {
+                // for non-default constructible types we also provide
+                // a customization point to save any external data
+                // (in the same way as Boost.Serialization does)
+                for (const auto& i: collection)
+                {
+                    save_construct_data(ar, &i, 0);
+                    ar << i;
+                }
+            }
+        };
+
+    public:
+        using type = typename std::conditional<
+            std::is_default_constructible<Value>::value,
+                default_, non_default_>::type;
+    };
+
+    template <class Value>
+    class load_collection_impl
+    {
+        struct default_
+        {
+            template <class Archive, class Collection>
+            static void call(Archive& ar, Collection& collection,
+                    typename Collection::size_type size)
+            {
+                collection.resize(size);
+                for (auto& i: collection)
+                    ar >> i;
+            }
+        };
+
+        struct non_default_
+        {
+            template <class Archive, class Collection>
+            static void call(Archive& ar, Collection& collection,
+                    typename Collection::size_type size)
+            {
+                using value_type = typename Collection::value_type;
+                using storage_type = std::aligned_storage<
+                    sizeof(value_type), alignof(value_type)>;
+
+                collection.clear();
+                while (size-- > 0)
```

Comment:
Would it be possible to call `collection.reserve(size);` here (at least for `vector` and `deque` etc.)?